### DISCUSSION TITLE
Extract the metadata title from a header

### DISCUSSION THREAD

**Discussion by 4l1fe** [getpelican/pelican#3290] [Q&A]
Hey. Most of the time i name files shortly and write the fuul title inside a file content and would like a page have slug and title extracted from it, it's like `<h1>` in html

Is it possible to set up?

**Comment by plgruener** [comment]
What format do you write your pages in? If markdown/rst, you should rather just use the :title: metadata field (inside each file) and adjust your template to automatically use that as the page title/ format it as `<h1>{{page.title}}</h1>` or something. You may also need to turn off the 'get title from filename' option.   
If you write directly in html, I think there's also option to embed metadata in there (via `<meta>` tags, but not sure).

**Reply by mcepl** [reply]
Well, except with rST this works:

```rst
Harry and Ginny as siblings
###########################

:date: 2019-07-15T23:07:27
:category: literature
:tags: storyPrompt, harryPotter, plotBunny, fromReddit

Originally published as “`Harry & Ginny are brother and sister`_”
on Reddit.
```

and the post has both slug, and metadata title generated from that rST headline.
https://matej.ceplovi.cz/blog/harry-and-ginny-as-siblings.html

Except if I try to do the same with Markdown:

```markdown
# Harry and Ginny as siblings

Date: 2019-07-15T23:07:27
Category: literature
Tags: storyPrompt, harryPotter, plotBunny, fromReddit

Originally published as "[Harry & Ginny are brother and
sister](https://www.reddit.com/r/HPfanfiction/comments/bevfuu/lf_harry_ginny_are_brother_and_sister/)"
on Reddit.
```

I get an error:

```
/archiv/knihovna/blog/publishconf.py -t "/home/matej/repos/pelican-themes/zurb-F5-basic"
[00:11:44] ERROR    Skipping                                                               contents.py:176
                    /home/matej/archiv/knihovna/blog/content/literature/hinny_siblings.md:                
                    could not find information about 'title'
```

**Comment by justinmayer** [comment]
I do not use reStructuredText enough to understand why it supports this use case, but the [code makes it clear that every page must have a title](https://github.com/getpelican/pelican/blob/3d3265b74c54338b8e9371c51eba4e8ef8b629c9/pelican/contents.py#L607), and the aforementioned [error is raised when mandatory metadata is missing](https://github.com/getpelican/pelican/blob/3d3265b74c54338b8e9371c51eba4e8ef8b629c9/pelican/contents.py#L176).

I think perhaps a `HEADING_METADATA` setting could be a way to add support for this use case, where if `Title` metadata is not explicitly defined, maybe setting `HEADING_METADATA = 0` uses the first heading as the title, a value of `-1` uses the last heading as the title, and `HEADING_METADATA = None` (default) would keep the current behavior. Just an idea.

Setting aside the implementation details, I think this might be a good candidate as a Pelican plugin, for whomever would like to add support for this use case.

**Reply by mcepl** [reply]
See https://github.com/getpelican/pelican/pull/3539 what I have vibe-coded on the topic.
